src/main/java/org/broadinstitute/consent/http/resources/DataAccessRequestResource.java

```diff
@@ -468,4 +475,35 @@ private User findUserByEmail(String email) {
         }
         return user;
     }
+
+    private DataAccessRequest findDataAccessRequestById(String referenceId) {
+        DataAccessRequest dar =  dataAccessRequestService.findByReferenceId(referenceId);
+        if (Objects.nonNull(dar)) {
+            return dar;
+        }
+        throw new NotFoundException("Unable to find Data Access Request with the provided reference id: " + referenceId);
+    }
+
+    /**
+     * Custom handler for validating that a user can access a DAR. User will have access if ANY
+     * of these conditions are met:
+     *      If the DAR create user is the same as the Auth User, then the user can access the resource.
+     *      If the user has any of the roles in allowableRoles, then the user can access the resource.
+     * In practice, pass in allowableRoles for users that are not the create user (i.e. Admin) so
+     * they can also have access to the DAR.
+     *
+     * @param allowableRoles List of roles that would allow the user to access the resource
+     * @param authUser The AuthUser
+     * @param referenceId The referenceId of the resource.
+     */
+    private void validateAuthedRoleUser(final List<UserRoles> allowableRoles, AuthUser authUser, String referenceId) {
+        DataAccessRequest dataAccessRequest = findDataAccessRequestById(referenceId);
+        User user = findUserByEmail(authUser.getName());
+        if (Objects.nonNull(dataAccessRequest.getUserId()) && dataAccessRequest.getUserId() > 0) {
```

Comment:
Don't know much about how the Java back-end is set up, so I have some questions. Is userId the only way to validate the user before role check? Is there a session token or timed OAuth token saved in server memory or db that can be referenced?

Reply:
The `AuthUser authUser` is the OAuthed user that we get from the bearer token when the user makes an authenticated request, so that is, in effect, what you're asking in this question, I believe. That information is validated against our user table in `findUserByEmail(authUser.getName())` so we ensure that the person who logged in is really a user we know about.

Reply:
And the AuthUser is determined at log in right? There's no way to alter that on each request?

Reply:
That's exactly right. If you look at the API methods, they are instantiated with `@Auth AuthUser authUser` which is populated by the framework. There are auth filters in place that examine the bearer token, query Google for the user, and then populate the `AuthUser`.

Reply:
Ok and validateAuthedRoleUser call ensures that a request can't impersonate another user by making sure that the userId provided in the request matches with the AuthUser id on the server right?

Reply:
Correct. Here is that method in `Resource`: 
```
    /**
     * Validate that the current authenticated user can access this resource.
     * If the user has one of the provided roles, then access is allowed.
     * If not, then the authenticated user must have the same identity as the
     * `userId` parameter they are requesting information for.
     *
     * @param authedRoles   List of UserRoles enums
     * @param authedUser    The authenticated DACUser
     * @param userId        The id of the DACUser the authenticated user is requesting access to
     */
    void validateAuthedRoleUser(final List<UserRoles> authedRoles, final User authedUser, final Integer userId) {
        List<Integer> authedRoleIds = authedRoles.stream().
                map(UserRoles::getRoleId).
                collect(Collectors.toList());
        boolean authedUserHasRole = authedUser.getRoles().stream().
                anyMatch(userRole -> authedRoleIds.contains(userRole.getRoleId()));
        if (!authedUserHasRole && !authedUser.getDacUserId().equals(userId)) {
            throw new ForbiddenException("User does not have permission");
        }
    }
```